### PR TITLE
chore(cu29_clock): guard against a zero time delta and potential overflow

### PR DESCRIPTION
**Pull request by makeecat**
## Summary

The calibration now guards against a zero time delta and potential overflow when computing the frequency, avoiding divide-by-zero and saturating to `u64::MAX` on overflow
## Changes
In `core/cu29_clock/src/calibration.rs`:
  - calibrate now requires both `counter_diff > 0` and a nonzero `time_diff`_ns via `NonZeroU128::new(...)`.
  - Frequency calculation is promoted to `u128`, then converted to `u64` with `try_from`, falling back to `u64::MAX` on overflow.

## Testing
- [x] `just std-ci`
- [x] `just lint`
- [x] `cargo +stable nextest run --workspace --all-targets`
- [x] `just nostd-ci`

## Checklist
- [x] I have updated docs or examples where needed
- [x] I have added or updated tests where needed
- [x] I have considered platform impact (Linux/macOS/Windows/embedded)
- [x] I have considered config/logging changes (if applicable)
- [x] This change is not a breaking change (or I documented it below)

## Additional context
This change reduces the risk of runtime panics in the clock calibration path. A zero time delta can occur due to coarse timer resolution, jitter, or identical timestamp sampling; previously, that would have triggered a divide-by-zero. The new guard prevents a crash and ensures deterministic, bounded behavior by saturating the frequency when overflow would occur. That’s important for predictable runtime behavior and reliable replay in deterministic systems.

### CODE REVIEW

File: core/cu29_clock/src/calibration.rs [hunk 1]
```diff
@@ -24,8 +24,11 @@ pub fn calibrate(
     let counter_diff = end_counter.saturating_sub(start_counter);
     let time_diff_ns = end_time.saturating_sub(start_time);
 
-    if counter_diff > 0 {
-        let freq_ns = ((counter_diff as u128 * 1_000_000_000) / time_diff_ns as u128) as u64;
+    if counter_diff > 0
+        && let Some(time_diff) = core::num::NonZeroU128::new(u128::from(time_diff_ns))
+    {
+        let freq_ns_u128 = (u128::from(counter_diff) * 1_000_000_000u128) / time_diff.get();
+        let freq_ns = u64::try_from(freq_ns_u128).unwrap_or(u64::MAX);
```
[the remark this answers is not in the document]
Comment (makeecat) [reply core/cu29_clock/src/calibration.rs]: this is tricky. Seems to have three options, open for discussion

  1. Option A: Minimal guard only. Steps: add a freq == 0 guard in counter_to_nanos to return init_time_ns (or 0) instead of dividing; add a unit test that calls counter_to_nanos before any successful calibration and asserts no panic and stable output. Pros: smallest change, lowest risk. Cons: time can freeze until a successful calibration, calibration failures are silent.
  2. Option B: Retry calibration + safe fallback. Steps: if time_diff_ns == 0 or counter_diff == 0, retry calibration up to N times with a slightly longer sleep; if still zero and FREQUENCY_NS is 0, set a nonzero fallback (e.g., 1_000_000_000) and refresh INIT_* to current values; keep a freq == 0 guard in counter_to_nanos as last resort; add tests for the zero‑delta path and fallback. Pros: avoids panic and time continues progressing. Cons: possible accuracy loss on true cycle counters; slightly more complexity.
  3. Option C: Explicit failure path + alternate time source. Steps: change calibrate to return success/failure and store that state; if calibration fails, CuInstant::now() falls back to RTC time (std) or returns a deterministic “not calibrated” path; add tests around failure behavior. Pros: explicit and debuggable, no silent incorrect scaling. Cons: more invasive API/behavior change, especially for no‑std.
  4. We ignore this, seems okay to me as well

[the remark this answers is not in the document]
Comment (gbin) [reply core/cu29_clock/src/calibration.rs]: Let's step back here... 
first how can it be 0? 
only the if the RTC clock is broken or SO bad we cannot really calibrate on it I would say.
When this thing happens: once, at boot time so if it crashes, no harm is done.

So for me this is the limit of the defensive programming: we don't want to handle anything here, let it crash this is the best behaviour.

**Review by gbin — CHANGES_REQUESTED** [review]
see comment, I would not bother. Defensive programming is only good if the outcome should not be a crash, here I believe this should crash. (but if I miss something feel free to comment)

File: core/cu29_clock/src/calibration.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (makeecat) [reply core/cu29_clock/src/calibration.rs]: Exactly, it will happen when the RTC clock is broken. RTC clock can happen during the execution due to hardware failure.

I agree that: let it crash is the best behavior. We should probably add a comment NOTE there for future reference?

[the remark this answers is not in the document]
Comment (gbin) [reply core/cu29_clock/src/calibration.rs]: good call for some users with a broken RTC they will have a stacktrace pointing to the comment saying: "if you get a div by zero here, check your hardware"

**Review by gbin — APPROVED** [review]
